packages/legend-query/src/stores/QueryBuilderMilestoningHelper.ts

```diff
@@ -0,0 +1,641 @@
+        propertyExpression.func.genericType.value.rawType,
+        graph,
+      );
+
+      if (targetStereotype) {
+        const sourceStereotype = getSourceTemporalStereotype(
+          guaranteeType(func, DerivedProperty),
+          graph,
+        );
+        if (
+          sourceStereotype !== MILESTONING_STEROTYPE.BITEMPORAL &&
+          targetStereotype !== sourceStereotype
+        ) {
+          if (targetStereotype === MILESTONING_STEROTYPE.BITEMPORAL) {
+            if (
+              propertyExpression.parametersValues.length !== 3 &&
+              !sourceStereotype
+            ) {
+              throw new UnsupportedOperationError(
+                "Property of milestoning sterotype 'Bitemporal' should have two parameters",
+              );
+            } else if (propertyExpression.parametersValues.length < 2) {
+              throw new UnsupportedOperationError(
+                "Property of milestoning sterotype 'Bitemporal' should have atleast one parameter",
+              );
+            } else if (propertyExpression.parametersValues.length > 3) {
+              throw new UnsupportedOperationError(
+                "Property of milestoning sterotype 'Bitemporal' should not have more than two parameters",
+              );
+            }
+          } else if (propertyExpression.parametersValues.length !== 2) {
+            throw new UnsupportedOperationError(
+              `Property of milestoning sterotype '${targetStereotype}' should have one parameters`,
+            );
+          }
+        }
+      }
+    }
+  }
+};
+
+// Gets the value of milestoning date that needs to be shown in DerivedPropertyEditor when date propagation is supported
+export const getPropagatedDate = (
+  derivedPropertyExpressionState: QueryBuilderDerivedPropertyExpressionState,
+  idx: number,
+): ValueSpecification | undefined => {
+  const queryBuilderState = derivedPropertyExpressionState.queryBuilderState;
+  const targetStereotype = getMilestoneTemporalStereotype(
+    guaranteeType(
+      derivedPropertyExpressionState.derivedProperty.genericType.value.rawType,
+      Class,
+    ),
+    queryBuilderState.graphManagerState.graph,
+  );
+  switch (targetStereotype) {
+    case MILESTONING_STEROTYPE.BITEMPORAL:
+      if (idx === 0) {
+        return queryBuilderState.querySetupState.ProcessingDate;
+      } else {
+        return queryBuilderState.querySetupState.businessDate;
+      }
+    case MILESTONING_STEROTYPE.BUSINESS_TEMPORAL:
+      return queryBuilderState.querySetupState.BusinessDate;
+    case MILESTONING_STEROTYPE.PROCESSING_TEMPORAL:
+      return queryBuilderState.querySetupState.processingDate;
+    default:
+      return undefined;
+  }
+};
+
+// To populate default dates to the next level in the property chain if the date values of the current
+// level are changed when date propagation is supported to the next level.
+export const propagateDefaultDates = (
```

Comment:
Let's dicuss about this. I wonder if this would be changed when we move to use `PropagatedMilestoningParameterValue` type instead